src/schema/profileAutocomplete.ts

```diff
@@ -0,0 +1,248 @@
+    name: String!
+  }
+
+  union AutocompleteHit = ExperienceHit | CompanyHit | SkillHit
+
+  type AutocompleteResult {
+    query: String!
+    limit: Int
+    hits: [AutocompleteHit!]!
+  }
+
+  extend type Query {
+    """
+    Get autocomplete suggestions for various fields like job titles, companies, skills, etc.
+    """
+    profileAutocomplete(
+      type: String!
+      query: String!
+      limit: Int
+    ): AutocompleteResult!
+  }
+`;
+
+export const DEFAULT_AUTOCOMPLETE_LIMIT = 10;
+
+const profileAutocompleteSchema = z.object({
+  type: z.nativeEnum(AutocompleteType, {
+    errorMap: () => ({ message: 'Invalid autocomplete type' }),
+  }),
+  query: z
+    .string()
+    .min(2, 'Query must be at least 2 characters long')
+    .max(100, 'Query must not exceed 100 characters'),
+  limit: z
+    .number()
+    .int()
+    .positive()
+    .optional()
+    .default(DEFAULT_AUTOCOMPLETE_LIMIT),
+});
+
+// Type inference from the schema
+type ProfileAutocompleteInput = z.infer<typeof profileAutocompleteSchema>;
+
+export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
+  unknown,
+  BaseContext
+>({
+  Query: {
+    profileAutocomplete: async (
+      _,
+      params: ProfileAutocompleteInput,
+      ctx: Context,
+    ) => {
+      const validation = profileAutocompleteSchema.safeParse(params);
+
+      if (!validation.success) {
+        if (validation.error.formErrors.fieldErrors.query) {
+          return {
+            query: params.query ?? '',
+            limit: params.limit ?? DEFAULT_AUTOCOMPLETE_LIMIT,
+            hits: [],
+          };
+        }
+
+        throw new ValidationError(validation.error.message);
+      }
+
+      const { type, query, limit }: ProfileAutocompleteInput = validation.data;
+      const {
+        entity,
+        propertyName,
+        where = {},
+        select,
+      } = autocompleteQueryMap[type];
+
+      const typeNameByEntity = {
+        [UserWorkExperience.name]: 'ExperienceHit',
+        [Company.name]: 'CompanyHit',
+        [UserSkill.name]: 'SkillHit',
+        [UserCertificationExperience.name]: 'ExperienceHit',
+        [UserAwardExperience.name]: 'ExperienceHit',
+        [UserPublicationExperience.name]: 'ExperienceHit',
+        [UserCourseExperience.name]: 'ExperienceHit',
+        [UserEducationExperience.name]: 'ExperienceHit',
+      } as const;
+
+      const hits = await queryReadReplica(ctx.con, ({ queryRunner }) =>
+        queryRunner.manager
+          .getRepository(entity)
+          .createQueryBuilder('entity')
+          .select(select ?? ['id', propertyName])
+          .where({
+            [propertyName]: Raw(
+              (alias) => `LOWER(${alias}) LIKE LOWER(:query)`,
+              { query: `%${query}%` },
+            ),
+            // extending where condition using the autocompleteQueryMap
+            ...where,
+          })
+          // sort by selected property name
+          .orderBy(`entity.${propertyName}`, 'ASC')
```

Comment:
do we have order by index?

Reply:
Ok, with this kind of dynamic column building its harder to see what columns are used from code, so that is why I asked. For example you change the mappings above but forget to add index. 

If it was just `orderBy('createdAt', 'ASC')` (just an example) you immediately know what is used.